doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,617 @@
+
+Options:
+
+  - Require applications to be changed to not use QUIC with a `BIO_f_buffer`.
+  - Detect when a `BIO_f_buffer` is part of a BIO stack and bypass it
+    (yucky and surprising).
+
+#### MTU Signalling
+
+**See also:**
+[BIO_s_dgram_pair(3)](https://www.openssl.org/docs/manmaster/man3/BIO_s_dgram_pair.html)
+
+`BIO_dgram_get_mtu` (`BIO_CTRL_DGRAM_GET_MTU`) and `BIO_dgram_set_mtu`
+(`BIO_CTRL_DGRAM_SET_MTU`) already exist for `BIO_s_dgram` and are implemented
+on a `BIO_s_dgram_pair` to allow the MTU to be determined and configured. One
+side of a pair can configure the MTU to allow the other side to detect it.
+
+`BIO_s_dgram` also has pre-existing support for getting the correct MTU value
+from the OS using `BIO_CTRL_DGRAM_QUERY_MTU`.
+
+### New APIs
+
+#### `BIO_sendmmsg` and `BIO_recvmmsg`
+
+**See also:**
+[BIO_sendmmsg(3)](https://www.openssl.org/docs/manmaster/man3/BIO_sendmmsg.html)
+
+The BIO interface features a new high-performance API for the execution of
+multiple read or write operations in a single system call, on supported OSes. On
+other OSes, a compatible fallback implementation is used.
+
+Unlike all other BIO APIs, this API is intended for concurrent threaded use and
+as such operates in a stateless fashion with regards to a BIO. This means, for
+example, that retry indications are made using explicit API inputs and outputs
+rather than setting an internal flag on the BIO.
+
+This new BIO API includes:
+
+- Local address support (getting the destination address of an incoming
+  packet; setting the source address of an outgoing packet), where support
+  for this is available;
+- Peer address support (setting the destination address of an outgoing
+  packet; getting the source address of an incoming packet), where support
+  for this is available.
+
+The following functionality was intentionally left out of this design because
+not all OSes can provide support:
+
+- Iovecs (which have also been determined not to be necessary for a
+  performant QUIC implementation);
+- Features such as `MSG_DONTWAIT`, etc.
+
+This BIO API is intended to be extensible. For more information on this API, see
+BIO_sendmmsg(3) and BIO_recvmmsg(3).
+
+Custom BIO implementers may set their own implementation of these APIs via
+corresponding `BIO_meth` getter/setter functions.
+
+#### Truncation Mode
+
+**See also:**
+[BIO_s_dgram_pair(3)](https://www.openssl.org/docs/manmaster/man3/BIO_s_dgram_pair.html)
+
+The controls `BIO_dgram_get_no_trunc` (`BIO_CTRL_DGRAM_GET_NO_TRUNC`) and
+`BIO_dgram_get_no_trunc` (`BIO_CTRL_DGRAM_GET_NO_TRUNC`) are introduced. This is
+a boolean value which may be implemented by BIOs with datagram semantics. When
+enabled, attempting to receive a datagram such that the datagram would
+ordinarily be truncated (as per the design of the Berkeley sockets API) instead
+results in a failure. This is intended for implementation by `BIO_s_dgram_pair`.
+For compatibility, the default behaviour is off.
```

Comment:
But we're talking about a BIO implementing datagrams not QUIC here. This is a feature that the QUIC implementation could use to detect accidental truncations and react accordingly (increase the MTU perhaps and our internal buffer sizes???)

Reply:
This still isn't making sense to me. Are you saying that the BIO might be MTU aware and know when truncation is happening when we're sending a packet? I'm not sure how a BIO is supposed to detect truncation.

Reply:
> Are you saying that the BIO might be MTU aware and know when truncation is happening

Yes

> when we're sending a packet?

I think this more about when we are *receiving* a packet, i.e. the packet is longer than the buffer we have allocated for it and it gets truncated when we read it.

> I'm not sure how a BIO is supposed to detect truncation.

Note that the document says "This is intended for implementation by `BIO_s_dgram_pair`". Effectively `BIO_s_dgram_pair` is like a mem BIO and so knows the length of the packet to be received, and the length of the buffer from the caller that its been provided with, and therefore is able to detect truncation.

Reply:
@kroeckx This isn't for application-side use. It's for the BIO that connects the QUIC implementation to the actual network (or some memory-based BIO pretending to be a network).

Basically what this comes down to is that the Berkeley sockets API is kind of crummy and not really very suited to use with UDP. In particular, the Berkeley sockets API, when receiving from a `SOCK_DGRAM` socket, will silently truncate the datagram if the buffer provided isn't big enough. I hate this, but unfortunately it's how it is.

These semantics are the default for `BIO_s_dgram_pair` because a) it's how `BIO_s_dgram` has to be implemented since it uses the OS APIs, b) basically everyone is probably going to expect a pipe with datagram semantics to behave like this. But since randomly truncating datagrams is not actually a good design, I've given it the option to be disabled.

Reply:
We might defer the implementation of these controls after MVP.